Breakpoint Mixin `max-width` Values

millllllz opened the pull request:
# Why?
Browser windows can be fractions of pixels wide, under certain circumstances (e.g. zoom). In these cases, it is not enough to set the `max-width` to the `em` equivalent of one pixel less.

# What changed?
* updated `max-width` calculation
* updated tests to match new output

# Notes
I recognize this is nearly moving the goalposts and doesn't solve the core problem, but in lieu of a exclusive max-width, 4 places of accuracy seems to be the standard and should work for most.

millllllz committed:
decrement bp-max by 0.0001 rather than 1/16

Browser windows can be fractions of pixels wide, under certain circumstances (e.g. zoom). In these cases, it is not enough to set the max-width to the em equivalent of one pixel less.

millllllz committed:
update max-widths values

ncoden commented:
This should fix:
* https://github.com/zurb/foundation-sites/issues/10820 show-for() and hide-for() gap at exact width of breakpoint

Related to:
* https://github.com/zurb/foundation-sites/issues/10808 XY-Grid Container doesn't fit next to each other in Edge

ncoden commented on scss/util/_breakpoint.scss:
```diff
@@ -77,7 +77,7 @@ $breakpoint-classes: (small medium large) !default;
   $bp: -zf-bp-to-em($bp);
   @if $bp-max {
-    $bp-max: -zf-bp-to-em($bp-max) - (1/16);
+    $bp-max: -zf-bp-to-em($bp-max) - 0.0001;
```
Why `0.0001` ?

Side note, numbers < 1 are usually written without the first `0` in CSS/Sass. Like `.0001`

JasonMiller replied:
Since all the `em` math works in 16ths, values end up at 4 decimals. I can remove the leading zero. Do you have a suggestion other than `.0001`?

ncoden commented on test/sass/_breakpoint.scss:
```diff
@@ -24,10 +24,10 @@
   @include test('Breakpoint (Only Range) [function]') {
     $test: breakpoint(medium only);
-    $expect: '(min-width: 40em) and (max-width: 63.9375em)';
+    $expect: '(min-width: 40em) and (max-width: 63.9999em)';
```
There is some tests to do in IE/Edge to ensure this works properly (or at least better than now). See #10808

JasonMiller replied:
Looking at #10808, wouldn't `.9375` be just as much a problem as `.9999`? I don't see this as introducing new issues. Thoughts?

ncoden replied:
True but I prefer testing. Also, is there any reason for `.9999` and not `.999999` ?

JasonMiller replied:
This is from my [previous comment](https://github.com/zurb/foundation-sites/pull/10978#discussion_r170259770), but since all the em math works in 16ths, values end up at 4 decimals. Do you have a suggestion other than `.0001` which would result in `.9999`?

ncoden replied:
According to the W3C CSS3 reference:

> CSS theoretically supports infinite precision and infinite ranges for all value types; however in reality implementations have finite capacity. UAs should support reasonably useful ranges and precisions.
> -- https://www.w3.org/TR/css-values-3/#numeric-types

It is common is CSS to write values with 6 decimals as it is precise enough to ensure a sub-pixel difference. But when dealing with breakpoints I'm not sure it's what we want.

JasonMiller replied:
I understand your concern. What can I do to move this along?

JasonMiller replied:
@ncoden is there a testing procedure you'd like me to follow here?

ncoden replied:
@JasonMiller I just have some tests to do across browsers, when I have the time for.

ncoden reviewed (commented):
Thank you @JasonMiller. There is some tests and investigations to do (like why did we used `1px` in the first place and not lower), but it's a nice catch.

See also the remarks below.

millllllz committed:
remove leading zero from bp-max decrement

JasonMiller commented:
@ncoden, I'm seeing TravisCI failures that do not pertain to my deltas. Please advise.

ncoden commented:
@JasonMiller We know about the unit test fails. We are resolving it.

ncoden commented:
@JasonMiller I just have some tests to do across browsers, when I have the time for.

DanielRuf commented:
What is the current state? The change would solve calculation bugs in Firefox and other browsers.

ncoden commented:
The current state was "waiting for @ncoden to remember this PR". Thanks for the recall.

DanielRuf commented:
related issues
https://github.com/zurb/foundation-sites/issues/10779
https://github.com/zurb/foundation-sites/issues/10695

ncoden committed:
fix: use 6-decimal precision between breakpoint

6 decimals is an arbitrary choice. Big enough to give sub-pixels renderd widths in mose use cases and small enough to not being round by browsers.

ncoden commented:
I tested it across browsers (Edge 16 and IE 9), it works fine and does not introduce new bugs. I also changed the precision from 4 to 6 decimals like discussed before, and applied the same precision to visibility classes.

ncoden committed:
fix: revert to Sass default 5-digit precision for breakpoint gutters

I was wrong, Sass actually has a default precision of 5 digits. Bootstrap-sass uses 8 digits but require to update the Sass compiler options.
See: https://github.com/twbs/bootstrap-sass#sass-number-precision

Note: this also update breakpoint tests

ncoden commented:
> I was wrong, Sass actually has a default precision of 5 digits. Bootstrap-sass uses 8 digits but require to update the Sass compiler options.
> See: https://github.com/twbs/bootstrap-sass#sass-number-precision

I updated tests too.

ncoden commented:
Seems good so far. Thank you @JasonMiller for the PR 👍

JasonMiller commented:
@ncoden, sorry to re-open this, but I'm seeing some issues with Chrome at exact breakpoint values, potentially related to rounding. If you take a look at [this codepen](https://codepen.io/millllllz/pen/rKavaM), and resize your screen to `1024px`, you'll see the 4-digit precision rule is still a match. This would be a problem for rules that use the `down` keyword, e.g. `medium down`. Did you see any of this in your testing? If you can confirm it's an issue, I can open a new pull request.
